### ISSUE TITLE
Duplicate diagnostics when using `--message-format json` with `--all-targets`/`--bins --tests`

### ISSUE THREAD

**Issue by Leandros** [rust-lang/rust#143007]
### Summary

Unlike when running for human readable output, the `--message-format json` output doesn't dedupe the diagnostics for different targets.

It's caused by the `test` and `bin` target both emitting the same diagnostic.

For running automatic code checking in the background (e.g., via bacon in combination with bacon-ls, this results in duplicated diagnostics in the diagnostics window).

This behavior is also shared by `cargo check`.

### Reproducer

I created an empty crate, with only a `main.rs` with an error:

```rust
// main.rs
fn main() {
  error;
}
```

```toml
# Cargo.toml
[package]
name = "empty"
version = "0.1.0"
edition = "2021"

[dependencies]
```

I expected to see only a single diagnostic when running 

```
cargo clippy --all-targets --message-format json
```

Instead, I saw the same diagnostic printed twice:

```json
{"reason":"compiler-message","package_id":"path+file:///local/home/gerstarv/tmp/empty#0.1.0","manifest_path":"/local/home/gerstarv/tmp/empty/Cargo.toml","target":{"kind":["bin"],"crate_types":["bin"],"name":"empty","src_path":"/local/home/gerstarv/tmp/empty/src/main.rs","edition":"2021","doc":true,"doctest":false,"test":true},"message":{"rendered":"error[E0425]: cannot find value `error` in this scope\n --> src/main.rs:2:5\n  |\n2 |     error;\n  |     ^^^^^ not found in this scope\n\n","$message_type":"diagnostic","children":[],"code":{"code":"E0425","explanation":"An unresolved name was used.\n\nErroneous code examples:\n\n```compile_fail,E0425\nsomething_that_doesnt_exist::foo;\n// error: unresolved name `something_that_doesnt_exist::foo`\n\n// or:\n\ntrait Foo {\n    fn bar() {\n        Self; // error: unresolved name `Self`\n    }\n}\n\n// or:\n\nlet x = unknown_variable;  // error: unresolved name `unknown_variable`\n```\n\nPlease verify that the name wasn't misspelled and ensure that the\nidentifier being referred to is valid for the given situation. Example:\n\n```\nenum something_that_does_exist {\n    Foo,\n}\n```\n\nOr:\n\n```\nmod something_that_does_exist {\n    pub static foo : i32 = 0i32;\n}\n\nsomething_that_does_exist::foo; // ok!\n```\n\nOr:\n\n```\nlet unknown_variable = 12u32;\nlet x = unknown_variable; // ok!\n```\n\nIf the item is not defined in the current module, it must be imported using a\n`use` statement, like so:\n\n```\n# mod foo { pub fn bar() {} }\n# fn main() {\nuse foo::bar;\nbar();\n# }\n```\n\nIf the item you are importing is not defined in some super-module of the\ncurrent module, then it must also be declared as public (e.g., `pub fn`).\n"},"level":"error","message":"cannot find value `error` in this scope","spans":[{"byte_end":21,"byte_start":16,"column_end":10,"column_start":5,"expansion":null,"file_name":"src/main.rs","is_primary":true,"label":"not found in this scope","line_end":2,"line_start":2,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":10,"highlight_start":5,"text":"    error;"}]}]}}
{"reason":"compiler-message","package_id":"path+file:///local/home/gerstarv/tmp/empty#0.1.0","manifest_path":"/local/home/gerstarv/tmp/empty/Cargo.toml","target":{"kind":["bin"],"crate_types":["bin"],"name":"empty","src_path":"/local/home/gerstarv/tmp/empty/src/main.rs","edition":"2021","doc":true,"doctest":false,"test":true},"message":{"rendered":"error[E0425]: cannot find value `error` in this scope\n --> src/main.rs:2:5\n  |\n2 |     error;\n  |     ^^^^^ not found in this scope\n\n","$message_type":"diagnostic","children":[],"code":{"code":"E0425","explanation":"An unresolved name was used.\n\nErroneous code examples:\n\n```compile_fail,E0425\nsomething_that_doesnt_exist::foo;\n// error: unresolved name `something_that_doesnt_exist::foo`\n\n// or:\n\ntrait Foo {\n    fn bar() {\n        Self; // error: unresolved name `Self`\n    }\n}\n\n// or:\n\nlet x = unknown_variable;  // error: unresolved name `unknown_variable`\n```\n\nPlease verify that the name wasn't misspelled and ensure that the\nidentifier being referred to is valid for the given situation. Example:\n\n```\nenum something_that_does_exist {\n    Foo,\n}\n```\n\nOr:\n\n```\nmod something_that_does_exist {\n    pub static foo : i32 = 0i32;\n}\n\nsomething_that_does_exist::foo; // ok!\n```\n\nOr:\n\n```\nlet unknown_variable = 12u32;\nlet x = unknown_variable; // ok!\n```\n\nIf the item is not defined in the current module, it must be imported using a\n`use` statement, like so:\n\n```\n# mod foo { pub fn bar() {} }\n# fn main() {\nuse foo::bar;\nbar();\n# }\n```\n\nIf the item you are importing is not defined in some super-module of the\ncurrent module, then it must also be declared as public (e.g., `pub fn`).\n"},"level":"error","message":"cannot find value `error` in this scope","spans":[{"byte_end":21,"byte_start":16,"column_end":10,"column_start":5,"expansion":null,"file_name":"src/main.rs","is_primary":true,"label":"not found in this scope","line_end":2,"line_start":2,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":10,"highlight_start":5,"text":"    error;"}]}]}}
{"reason":"compiler-message","package_id":"path+file:///local/home/gerstarv/tmp/empty#0.1.0","manifest_path":"/local/home/gerstarv/tmp/empty/Cargo.toml","target":{"kind":["bin"],"crate_types":["bin"],"name":"empty","src_path":"/local/home/gerstarv/tmp/empty/src/main.rs","edition":"2021","doc":true,"doctest":false,"test":true},"message":{"rendered":"For more information about this error, try `rustc --explain E0425`.\n","$message_type":"diagnostic","children":[],"code":null,"level":"failure-note","message":"For more information about this error, try `rustc --explain E0425`.","spans":[]}}
{"reason":"compiler-message","package_id":"path+file:///local/home/gerstarv/tmp/empty#0.1.0","manifest_path":"/local/home/gerstarv/tmp/empty/Cargo.toml","target":{"kind":["bin"],"crate_types":["bin"],"name":"empty","src_path":"/local/home/gerstarv/tmp/empty/src/main.rs","edition":"2021","doc":true,"doctest":false,"test":true},"message":{"rendered":"For more information about this error, try `rustc --explain E0425`.\n","$message_type":"diagnostic","children":[],"code":null,"level":"failure-note","message":"For more information about this error, try `rustc --explain E0425`.","spans":[]}}
{"reason":"build-finished","success":false}
```

When running without `--message-format json`, the output is as expected:

```
    Checking empty v0.1.0 (/local/home/leandros/tmp/empty)
error[E0425]: cannot find value `error` in this scope
 --> src/main.rs:2:5
  |
2 |     error;
  |     ^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `empty` (bin "empty") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `empty` (bin "empty" test) due to 1 previous error
```

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: x86_64-unknown-linux-gnu
release: 1.84.0
LLVM version: 19.1.5
```

**Comment by blyxyas** [comment]
Oops, turns out that `--message-format <FMT>` is a Cargo-only flag, [handled here](https://github.com/rust-lang/cargo/blob/409fed7dc1553d49cb9a8c0637d12d65571346ce/src/cargo/ops/cargo_package/mod.rs#L284-L301).
